libp2p/module.go

```diff
@@ -0,0 +1,358 @@
+	"github.com/pokt-network/pocket/libp2p/protocol"
+	"github.com/pokt-network/pocket/logger"
+	"github.com/pokt-network/pocket/p2p/providers"
+	"github.com/pokt-network/pocket/p2p/providers/addrbook_provider"
+	"github.com/pokt-network/pocket/p2p/providers/addrbook_provider/persistence"
+	typesP2P "github.com/pokt-network/pocket/p2p/types"
+	"github.com/pokt-network/pocket/runtime/configs"
+	"github.com/pokt-network/pocket/runtime/configs/types"
+	poktCrypto "github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/messaging"
+	"github.com/pokt-network/pocket/shared/modules"
+)
+
+var _ modules.P2PModule = &libp2pModule{}
+
+type libp2pModule struct {
+	logger                *modules.Logger
+	bus                   modules.Bus
+	cfg                   *configs.P2PConfig
+	addrBookProvider      providers.AddrBookProvider
+	currentHeightProvider providers.CurrentHeightProvider
+	identity              libp2p.Option
+	listenAddrs           libp2p.Option
+	host                  host.Host
+	pubsub                *pubsub.PubSub
+	topic                 *pubsub.Topic
+	subscription          *pubsub.Subscription
+	network               typesP2P.Network
+}
+
+var (
+	// TODO: consider security exposure of and "safe minimum" for timeout.
+	// TODO: parameterize and expose via config.
+	// ReadStreamTimeout is the duration to wait for a read operation on a
+	// stream to complete, after which the stream is closed ("timed out").
+	readStreamTimeoutDuration = time.Second * 10
+	// ErrModule wraps errors which occur within the libp2pModule implementation.
+	ErrModule = typesP2P.NewErrFactory("LibP2P module error")
+)
+
+func Create(bus modules.Bus) (modules.Module, error) {
+	return CreateWithProviders(
+		bus,
+		persistence.NewPersistenceAddrBookProvider(bus),
+		bus.GetConsensusModule(),
+	)
+}
+
+func CreateWithProviders(
+	bus modules.Bus,
+	addrBookProvider addrbook_provider.AddrBookProvider,
+	currentHeightProvider providers.CurrentHeightProvider,
+) (modules.Module, error) {
+	return new(libp2pModule).CreateWithProviders(bus, addrBookProvider, currentHeightProvider)
+}
+
+func (mod *libp2pModule) GetModuleName() string {
+	// TODO: double check if this should change.
+	return modules.P2PModuleName
+}
+
+func (mod *libp2pModule) Create(bus modules.Bus) (modules.Module, error) {
+	return Create(bus)
+}
+
+func (mod *libp2pModule) CreateWithProviders(bus modules.Bus, addrBookProvider addrbook_provider.AddrBookProvider, currentHeightProvider providers.CurrentHeightProvider) (modules.Module, error) {
+	*mod = libp2pModule{
+		logger:                new(modules.Logger),
+		addrBookProvider:      addrBookProvider,
+		currentHeightProvider: currentHeightProvider,
+	}
+
+	if err := bus.RegisterModule(mod); err != nil {
+		return nil, ErrModule("unable to register module", err)
+	}
+
+	mod.logger.Print("Creating libp2p-backed network module")
+
+	mod.cfg = bus.GetRuntimeMgr().GetConfig().P2P
```

Comment:
This, and some others, can be moved directly into the struct initialisation

Reply:
I restructured things a bit, let me know if that's not what you had in mind.